packages/gatsby/src/utils/webpack/plugins/static-query-mapper.ts

```diff
@@ -76,390 +59,323 @@ function getRealPath(
   return cache.get(componentPath) as string
 }
 
-/**
- * Grab the actual webpackModule from the resourcePath
- * We return staticQueries and componentPaths cause that's what we care about
- */
-function getWebpackModulesByResourcePaths(
-  modules: Set<Module>,
-  staticQueries: IGatsbyState["staticQueryComponents"],
-  components: IGatsbyState["components"],
-  componentsUsingSlices: IGatsbyState["componentsUsingSlices"]
-): {
-  webpackModulesByStaticQueryId: Map<string, Set<Module>>
-  webpackModulesByComponentId: Map<string, Set<Module>>
-  webpackModulesUsingSlices: Set<{
-    module: Module
-    slices: Record<string, ICollectedSlice>
-  }>
-} {
-  const realPathCache = new Map<string, string>()
-  const webpackModulesByStaticQueryId = new Map<string, Set<Module>>()
-  const webpackModulesByComponentId = new Map<string, Set<Module>>()
-  const webpackModulesUsingSlices = new Set<{
-    module: Module
-    slices: Record<string, ICollectedSlice>
-  }>()
-
-  modules.forEach(webpackModule => {
-    for (const [id, staticQuery] of staticQueries) {
-      const staticQueryComponentPath = getRealPath(
-        realPathCache,
-        staticQuery.componentPath
-      )
-
-      if (
-        !doesModuleMatchResourcePath(staticQueryComponentPath, webpackModule)
-      ) {
-        continue
-      }
-
-      let set = webpackModulesByStaticQueryId.get(id)
-
-      if (!set) {
-        set = new Set()
-      }
+export class StaticQueryMapper {
+  private store: Store<IGatsbyState>
+  private name: string
 
-      set.add(webpackModule)
+  constructor(store) {
+    this.store = store
+    this.name = `StaticQueryMapper`
+  }
 
-      webpackModulesByStaticQueryId.set(id, set)
-    }
+  apply(compiler: Compiler): void {
+    const {
+      components,
+      staticQueryComponents,
+      componentsUsingSlices,
+      program,
+    } = this.store.getState()
 
-    for (const [id, component] of components) {
-      const componentComponentPath = getRealPath(
-        realPathCache,
-        component.componentPath
-      )
-      if (!doesModuleMatchResourcePath(componentComponentPath, webpackModule)) {
-        continue
+    compiler.hooks.done.tap(this.name, stats => {
+      // In dev mode we want to write page-data when compilation succeeds
+      if (!stats.hasErrors() && compiler.watchMode) {
+        enqueueFlush()
       }
+    })
 
-      let set = webpackModulesByComponentId.get(id)
+    compiler.hooks.finishMake.tapPromise(
+      {
+        name: this.name,
+        before: `PartialHydrationPlugin`,
```

Comment:
`before` here technically doesn't do anything yet, but in follow up PR, partial hydration plugin will use same step to modify moduleGraph and we want to handle static queries (and slices) before partial hydration plugin mutate moduleGraph